base/unicode/utf8.jl

```diff
@@ -312,6 +312,11 @@ function convert(::Type{UTF8String}, a::Vector{UInt8}, invalids_as::AbstractStri
 end
 convert(::Type{UTF8String}, s::AbstractString) = utf8(bytestring(s))
 
+function convert{T<:Union{UInt16, UInt32}}(::Type{UTF8String}, a::Vector{T}, len=length(a)-1)
+    encode_to_utf8(T, a, len)
```

Comment:
That length must be calculated by `checkstring`.  It needs to take into account characters that will take 2, 3, or 4 bytes when encoded as UTF8.
Also, `encode_to_utf8` does no validation, as it is intended ONLY to be used on strings that have passed `checkstring`'s validation.